Mathlib/Algebra/Hom/GroupAction.lean

```diff
@@ -0,0 +1,527 @@
+  coe_injective' f g h := by
+    rcases f with ⟨tF, _, _⟩; rcases g with ⟨tG, _, _⟩
+    cases tF; cases tG
+    simp only at h
+    simp only [h]
+  map_smul m := m.map_smul'
+  map_zero := DistribMulActionHom.map_zero'
+  map_add := DistribMulActionHom.map_add'
+
+variable {M A B}
+
+-- Porting note: Removed `to_fun_eq_coe` since it is now a syntactic tautology
+#noalign distrib_mul_action_hom.to_fun_eq_coe
+
+@[norm_cast]
+theorem coe_fn_coe (f : A →+[M] B) : ((f : A →+ B) : A → B) = f :=
+  rfl
+#align distrib_mul_action_hom.coe_fn_coe DistribMulActionHom.coe_fn_coe
+
+-- Porting note: Removed `coe_fn_coe'` since it is now a syntactic tautology
+#noalign distrib_mul_action_hom.coe_fn_coe'
+
+@[ext]
+theorem ext {f g : A →+[M] B} : (∀ x, f x = g x) → f = g :=
+  FunLike.ext f g
+#align distrib_mul_action_hom.ext DistribMulActionHom.ext
+
+theorem ext_iff {f g : A →+[M] B} : f = g ↔ ∀ x, f x = g x :=
+  FunLike.ext_iff
+#align distrib_mul_action_hom.ext_iff DistribMulActionHom.ext_iff
+
+protected theorem congr_fun {f g : A →+[M] B} (h : f = g) (x : A) : f x = g x :=
+  FunLike.congr_fun h _
+#align distrib_mul_action_hom.congr_fun DistribMulActionHom.congr_fun
+
+theorem toMulActionHom_injective {f g : A →+[M] B} (h : (f : A →[M] B) = (g : A →[M] B)) :
+    f = g := by
+  ext a
+  exact MulActionHom.congr_fun h a
+#align
+  distrib_mul_action_hom.to_mul_action_hom_injective DistribMulActionHom.toMulActionHom_injective
+
+theorem toAddMonoidHom_injective {f g : A →+[M] B} (h : (f : A →+ B) = (g : A →+ B)) : f = g :=
+  by
+  ext a
+  exact AddMonoidHom.congr_fun h a
+#align
+  distrib_mul_action_hom.to_add_monoid_hom_injective DistribMulActionHom.toAddMonoidHom_injective
+
+protected theorem map_zero (f : A →+[M] B) : f 0 = 0 :=
+  map_zero f
+#align distrib_mul_action_hom.map_zero DistribMulActionHom.map_zero
+
+protected theorem map_add (f : A →+[M] B) (x y : A) : f (x + y) = f x + f y :=
+  map_add f x y
+#align distrib_mul_action_hom.map_add DistribMulActionHom.map_add
+
+protected theorem map_neg (f : A' →+[M] B') (x : A') : f (-x) = -f x :=
+  map_neg f x
+#align distrib_mul_action_hom.map_neg DistribMulActionHom.map_neg
+
+protected theorem map_sub (f : A' →+[M] B') (x y : A') : f (x - y) = f x - f y :=
+  map_sub f x y
+#align distrib_mul_action_hom.map_sub DistribMulActionHom.map_sub
+
+protected theorem map_smul (f : A →+[M] B) (m : M) (x : A) : f (m • x) = m • f x :=
+  map_smul f m x
+#align distrib_mul_action_hom.map_smul DistribMulActionHom.map_smul
+
+variable (M)
+/-- The identity map as an equivariant additive monoid homomorphism. -/
+protected def id : A →+[M] A :=
+  ⟨.id _, rfl, fun _ _ => rfl⟩
+#align distrib_mul_action_hom.id DistribMulActionHom.id
+
+@[simp]
+theorem id_apply (x : A) : DistribMulActionHom.id M x = x := by
+  rfl
+#align distrib_mul_action_hom.id_apply DistribMulActionHom.id_apply
+
+variable {M C}
+
+-- porting note: `simp` used to prove this, but now `change` is needed to push past the coercions
```

Comment:
If `simp` cannot prove `↑↑0 = 0` then this is an indication that something is wrong, right?

Reply:
Yes, a bunch of `simp` lemmas were deleted for being syntactic tautologies when they should have been fixed to not be syntactic tautologies any more.

Reply:
I reintroduced the lemmas that #lint reported to be syntactic tautologies. To fix them, I laid out what seemed like a reasonable strategy [on Zulip.](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Algebra.2EHom.2EGroupAction/near/321253822) which is essentially what seems to happen in Algebra.Hom.Group as well. I could still be way off with that fix, so anyone who considers themselves an expert on coercion is welcome to polish this.

The `↑↑0 = 0` issue is not fixed by this, but I believe PR https://github.com/leanprover-community/mathlib4/pull/1219 is consistent with the changes I have made and will fix it by reintroducing OneHom.toFun_eq_coe (I mention this earlier in the Zulip thread).